### PR TITLE
Add HF benchmarks

### LINKED ISSUE #1716: Add HF models PEFT benchmarks
## 🚀 Feature

Add the following HF models with PEFT config to Thunder's benchmark suite:

- [x] `Qwen/Qwen2.5-7B-Instruct`
- [x] `microsoft/Phi-3-mini-128k-instruct`
- [x] `mistralai/Mistral-Nemo-Base-2407`
- [x] `bigcode/starcoder2-7b`

This is a tracking issue for the branch [`hf-nemo-benchmarks`](https://github.com/Lightning-AI/lightning-thunder/tree/hf-nemo-benchmarks).

cc @crcrpar @tfogal

### PR DESCRIPTION
**Pull request by riccardofelluga**
This PR adds the HF models benchmarks to Thunder's benchmarking suite for both `PEFT` and full training `FT`. It closes #1716. 

The added model ids are:
- `bigcode/starcoder2-7b`
- `google/gemma-7b`
- `google/gemma-2-9b-it`
- `meta-llama/Llama-3.2-3B`
- `meta-llama/Meta-Llama-3-8B-Instruct`
- `microsoft/Phi-3.5-mini-instruct`
- `microsoft/phi-4`
- `mistralai/Mistral-Nemo-Base-2407`
- `Qwen/Qwen2-7B`
- `Qwen/Qwen2.5-7B-Instruct`

Some of the models require HF token to access. 

ps. this PR is almost ready, I need to do the last checks so it's draft for now.

cc @crcrpar

### CODE REVIEW

**Comment by riccardofelluga** [comment]
No more draft! The only difference between this benchmarks and NeMo is in the LoRA layers which, in the case of NeMo are implemented using matumuls, whereas the HF peft library uses linear layers to achieve the same result.

NeMo implementation: https://github.com/NVIDIA/NeMo/blob/beab6d47d855eb51253f97dc6731e4417e373266/nemo/collections/llm/peft/lora.py#L143-L156 

HF peft implementation:
https://github.com/huggingface/peft/blob/2825774d2de1c8bd0604ac685867edd79d608a9e/src/peft/tuners/lora/layer.py#L709

To compare the two I added two more benchmarks to this suite that mock up the same model but uses the two different LoRA implementation. This can be tested with `pytest thunder/benchmarks/targets.py -k "lora_linear" --benchmark-timer=torch.utils.benchmark.utils.timer.timer --benchmark-warmup=on --benchmark-group-by=param:compute_type --benchmark-warmup-iterations=10`

Wierdly enough thunder seems to be very much slower than both inductor and eager: 

```shell
---------------------------------------------------------------------------- benchmark 'compute_type=ComputeType.TRAINING_BACKWARD': 6 tests -----------------------------------------------------------------------------
Name (time in us)                                  Min                   Max                Mean             StdDev              Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_nemo_lora_linear[backward-inductor]      279.9155 (1.0)      1,028.0015 (1.97)     307.5476 (1.0)      35.6273 (1.12)     300.4275 (1.0)       9.8018 (1.0)        88;108        3.2515 (1.0)        1705           2
test_hf_lora_linear[backward-inductor]        282.4720 (1.01)       521.4545 (1.0)      308.3400 (1.00)     31.7722 (1.0)      301.8200 (1.00)     10.4370 (1.06)       91;110        3.2432 (1.00)       1670           2
test_nemo_lora_linear[backward-eager]         308.9520 (1.10)     1,209.2750 (2.32)     329.3746 (1.07)     36.2003 (1.14)     322.4815 (1.07)     12.6656 (1.29)        82;86        3.0361 (0.93)       1569           2
test_hf_lora_linear[backward-eager]           328.8140 (1.17)     1,041.9055 (2.00)     358.5735 (1.17)     34.2569 (1.08)     353.2260 (1.18)      9.9289 (1.01)       82;120        2.7888 (0.86)       1478           2
test_hf_lora_linear[backward-thunderfx]       601.0530 (2.15)     2,034.0050 (3.90)     667.3585 (2.17)     94.1061 (2.96)     641.8175 (2.14)     29.9585 (3.06)      103;164        1.4984 (0.46)       1520           1
test_nemo_lora_linear[backward-thunderfx]     615.3560 (2.20)     1,146.2360 (2.20)     669.0339 (2.18)     85.2972 (2.68)     647.3741 (2.15)     24.5128 (2.50)       76;138        1.4947 (0.46)       1523           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------- benchmark 'compute_type=ComputeType.TRAINING_FORWARD': 6 tests -----------------------------------------------------------------------------
Name (time in us)                                 Min                   Max                Mean              StdDev              Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_nemo_lora_linear[forward-eager]         204.8484 (1.0)        359.6930 (1.0)      217.1435 (1.0)       26.8580 (1.0)      210.0090 (1.0)       4.4708 (1.0)       100;143        4.6052 (1.0)        1581           3
test_nemo_lora_linear[forward-inductor]      214.6163 (1.05)       391.2380 (1.09)     230.6942 (1.06)      29.3198 (1.09)     222.4005 (1.06)      6.2602 (1.40)       85;144        4.3347 (0.94)       1500           3
test_hf_lora_linear[forward-inductor]        229.4977 (1.12)       533.6100 (1.48)     248.2199 (1.14)      31.6476 (1.18)     239.2725 (1.14)      8.1190 (1.82)       85;133        4.0287 (0.87)       1410           3
test_hf_lora_linear[forward-eager]           266.8125 (1.30)     2,128.4015 (5.92)     287.4951 (1.32)      61.3338 (2.28)     274.2370 (1.31)      6.5187 (1.46)      114;172        3.4783 (0.76)       1807           2
test_hf_lora_linear[forward-thunderfx]       775.8250 (3.79)     2,610.0180 (7.26)     839.2712 (3.87)     113.6345 (4.23)     813.5395 (3.87)     25.9290 (5.80)       82;110        1.1915 (0.26)       1250           1
test_nemo_lora_linear[forward-thunderfx]     782.9630 (3.82)     1,231.3460 (3.42)     845.2695 (3.89)      91.7294 (3.42)     822.9025 (3.92)     27.6695 (6.19)       80;108        1.1831 (0.26)       1196           1
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

**Comment by IvanYashchuk** [comment]
If the only difference is in the LoRA implementation is it feasible to add another option to the benchmark to use the same implementation as in NeMo and copy it here?

**Comment by riccardofelluga** [comment]
> If the only difference is in the LoRA implementation is it feasible to add another option to the benchmark to use the same implementation as in NeMo and copy it here?

I think it can be done, I just need to make sure the extra method does not interfere with NeMo peft configuration process. I'll push an updated on that soon.

**Review by t-vi — APPROVED** [review]
Good to have this. Thank you @riccardofelluga and everyone.

File: thunder/benchmarks/__init__.py [hunk 1]
```diff
@@ -3013,6 +3013,163 @@ def fn(self) -> Callable:
...
+                    target_modules=["linear_proj", "linear_fc1", "linear_fc2"],
+                )
+                from peft import get_peft_model
+
+                # autocast_adapter_dtype set to False is needed to actually execute the ops in bf16
+                # otherwise they'll be casted to f32
+                self.lora_cls = partial(get_peft_model, peft_config=peft_config, autocast_adapter_dtype=False)
+
+            case "NeMo":
+                from nemo.collections.llm.peft import LoRA
```
Comment (IvanYashchuk) [inline thunder/benchmarks/__init__.py:3127]: Instead of importing this layer which requires nemo to be installed. Could you please copy the code for this layer to this benchmark implementation?

Comment (IvanYashchuk) [reply thunder/benchmarks/__init__.py:3127]: Is it "LinearLoRA" below?

Comment (riccardofelluga) [reply thunder/benchmarks/__init__.py:3127]: nemo  LoRA class is more like the config class in HF with some extra features. What could be done is to paste here the LinearAdapter layer from here https://github.com/NVIDIA/NeMo/blob/dc93e9df668f99be7e29ab4cc429e44e602ef220/nemo/collections/llm/peft/lora.py#L43

Comment (IvanYashchuk) [reply thunder/benchmarks/__init__.py:3127]: Let's copy and paste the code with a link in a comment where the code is taken from.
